### PR TITLE
Fixed #29019 -- Added ManyToManyField support to REQUIRED_FIELDS.

### PR DESCRIPTION
**Pull request by hramezani**
[ticket 29019](https://code.djangoproject.com/ticket/29019)

### CODE REVIEW

**Comment by hramezani** [comment]
@felixxm Sorry for the delay. 

> A helpful error message in the case of ManyToManyField with a through model is missing

Done.

**Comment by hramezani** [comment]
@felixxm anything else?

File: tests/auth_tests/models/with_many_to_many.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        user = self.model(username=username, email=email)
+        user.set_password(password)
+        user.save(using=self._db)
+
+        # Add organization
+        for organization_id in orgs.split():
+            try:
+                user.orgs.add(Organization.objects.get(id=int(organization_id)))
+            except Exception:
+                pass
```
Comment (felixxm) [inline tests/auth_tests/models/with_many_to_many.py]: `add()` accepts IDs so we can simplify this, e.g.
```python
...
user.save(using=self._db)
user.orgs.add(*orgs.split())
```

File: tests/auth_tests/models/with_many_to_many.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    email = models.EmailField()
+    orgs = models.ManyToManyField(Organization)
+
+    custom_objects = CustomUserWithM2MManager()
+
+    USERNAME_FIELD = 'username'
+    REQUIRED_FIELDS = ['email', 'orgs']
+
+
+class CustomUserWithM2MAndThrough(AbstractBaseUser):
```
Comment (felixxm) [inline tests/auth_tests/models/with_many_to_many.py]: `CustomUserWithM2MAndThrough` -> `CustomUserWithM2MThrough`

File: django/contrib/auth/management/commands/createsuperuser.py [hunk 3]
```diff
@@ -111,10 +121,11 @@ def handle(self, *args, **options):
                         message = self._get_input_message(field)
                         input_value = self.get_input_data(field, message)
                         user_data[field_name] = input_value
-                        fake_user_data[field_name] = input_value
+                        if not field.many_to_many:
+                            fake_user_data[field_name] = input_value
```
Comment (felixxm) [inline django/contrib/auth/management/commands/createsuperuser.py]: Comment above `fake_user_data = {}` should be updated.

File: tests/auth_tests/test_management.py [hunk 4]
```diff
@@ -500,6 +500,70 @@ def test(self):
...
+            CommandError,
+            "Required ManyToManyField 'orgs' has through model which is not supported."
+        ):
+            call_command(
+                'createsuperuser',
+                interactive=False,
+                username='joe',
+                email='joe@test.com',
+                orgs='1',
+                stdout=new_io,
```
Comment (felixxm) [inline tests/auth_tests/test_management.py]: `kwargs` are not necessary to raise an exception, e.g.
```python
@override_settings(AUTH_USER_MODEL='auth_tests.CustomUserWithM2MThrough')
def test_unsupport_fields_with_m2m_and_through(self):
    msg = (
        "Required field 'orgs' specifies a many-to-many relation through "
        "model, which is not supported."
    )
    with self.assertRaisesMessage(CommandError, msg):
        call_command('createsuperuser')
```

File: tests/auth_tests/test_management.py [hunk 5]
```diff
@@ -500,6 +500,70 @@ def test(self):
...
+        self.assertEqual(command_output, 'Superuser created successfully.')
+        u = CustomUserWithM2M._default_manager.get(username='joe')
+        self.assertEqual(u.email, 'joe@test.com')
+        self.assertEqual(u.orgs.count(), 2)
+
+    @override_settings(AUTH_USER_MODEL='auth_tests.CustomUserWithM2M')
+    def test_fields_with_m2m_interactive(self):
+        new_io = StringIO()
+        organization1 = Organization.objects.create(name='test organ1')
+        organization2 = Organization.objects.create(name='test organ2')
```
Comment (felixxm) [inline tests/auth_tests/test_management.py]: I would use `Organization` IDs here and in `test_fields_with_m2m()`, e.g.
```python
org_id_1 = Organization.objects.create(name='Organization 1').pk
org_id_2 = Organization.objects.create(name='Organization 2').pk
...
@mock_inputs({
    'password': 'nopasswd',
    'Username: ': 'joe',
    'Orgs (Organization.id): ': '%s %s' % (org_id_1, org_id_2),
})
```

File: tests/auth_tests/models/with_many_to_many.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+class Organization(models.Model):
+    name = models.CharField(max_length=255)
+
+
+class CustomUserWithM2MManager(BaseUserManager):
+    def create_superuser(self, username, email, orgs, password):
+        user = self.model(username=username, email=email)
+        user.set_password(password)
+        user.save(using=self._db)
+
```
Comment (felixxm) [inline tests/auth_tests/models/with_many_to_many.py]: Chop blank line.

File: tests/auth_tests/test_management.py [hunk 7]
```diff
@@ -500,6 +500,70 @@ def test(self):
...
+            'Orgs (Organization.id): ': ' '.join([str(organization1.pk), str(organization2.pk)]),
+        })
+        def test(self):
+            call_command(
+                'createsuperuser',
+                interactive=True,
+                stdout=new_io,
+                stdin=MockTTY(),
+            )
+
```
Comment (felixxm) [inline tests/auth_tests/test_management.py]: Chop blank line.

**Comment by felixxm** [comment]
Can we also add release notes?, e.g.:
```
* :attr:`~django.contrib.auth.models.CustomUser.REQUIRED_FIELDS` now supports
  :class:`~django.db.models.ManyToManyField`\s.
```
and a short note to the [REQUIRED_FIELDS documentation](https://docs.djangoproject.com/en/2.2/topics/auth/customizing/#django.contrib.auth.models.CustomUser.REQUIRED_FIELDS), e.g.
```
.. versionadded:: 3.0

    :attr:`REQUIRED_FIELDS` now supports
    :class:`~django.db.models.ManyToManyField`\s without a custom
    through model. Since there is no way to pass model instances during
    the :djadmin:`createsuperuser` prompt, expect the user to enter IDs
    of an existing instance of the class to which the model is related.
```

**Comment by hramezani** [comment]
@felixxm Thanks for the comments.
I did the requested changes in comments.
Also, I've added the documentation you provided regarding the release note.

**Comment by felixxm** [comment]
@hramezani After reconsideration I think it will be better to use a `list` option instead of passing a string to a manager. Can you take a look at my proposition? with this change you can use M2M fields multiple times, e.g. `--orgs=1 --orgs=2` and we'll pass a list to a manager.

**Comment by hramezani** [comment]
@felixxm Very good idea :+1: . Looks good to Me.

**Comment by hramezani** [comment]
@felixxm Thanks for awesome work on this PR and sorry for incomplete PR.

**Pull request merged**